File: joe/rock_paper_scissors.py
import sys
import random

# A tuple of tuples (computer_choice, user_choice) where the computer wins
WIN_CASES = (("rock", "scissor"), ("scissor", "paper"), ("paper", "rock"))
# Valid choices
CHOICES = ("rock", "paper", "scissors")

# Terminal escape codes for coloring text
GREEN_ANSI = "\033[92m"
END_ANSI = "\033[0m"

QUIT_KEYWORDS = (".quit", "q", "stop", "exit")

MESSAGE_WIN = (
    "🤯 You beat me, human. I chose {0} but you chose {1}. Now let's play again.\n"
)
MESSAGE_LOSE = "🚀 You got absolutely destroyed by the bigbrain computer, as I chose {0} versus your {1}. Wanna play again?\n"
MESSAGE_TIE = "🤔 Aw man, it was a tie... Let's play again.\n"

try:
    user_choice = ""
    while True:
        user_choice = input(
            f"{GREEN_ANSI}Rock, Paper, Scissors? Type your choice here or .quit to quit: {END_ANSI}"
        ).lower()

        if user_choice in QUIT_KEYWORDS:
            print("Shutting down the magical rock-paper-scissors AI...")
            sys.exit()
        if user_choice not in CHOICES:
            print("That wasn't a valid choice, please try again.")
            continue

        # Pick a random value
        bot_choice = random.choice(CHOICES)

        if (bot_choice, user_choice) in WIN_CASES:
            print(MESSAGE_LOSE.format(bot_choice, user_choice))
        elif bot_choice == user_choice:
            print(MESSAGE_TIE)
        else:
            print(MESSAGE_WIN.format(bot_choice, user_choice))

except KeyboardInterrupt:
    print("\nGood playing with you, human. I'll be shutting down now.")

File: joe/rock_paper_scissors.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strings"
	"time"
)

const (
	ChoiceRock    = "rock"
	ChoicePaper   = "paper"
	ChoiceScissor = "scissor"

	GreenAnsiStart = "\033[92m"
	AnsiEnd        = "\033[0m"
)

var (
	winCases = map[string]string{ChoiceRock: ChoiceScissor, ChoiceScissor: ChoicePaper, ChoicePaper: ChoiceRock}
	choices  = [3]string{ChoiceRock, ChoicePaper, ChoiceScissor}
)

func main() {
	rand.Seed(time.Now().UnixNano())

	for {
		choice := input("Rock, paper, scissors? ")

		switch choice {
		case ChoiceRock, ChoicePaper, ChoiceScissor:
			// ok
		case ".quit", "q", "stop", "exit":
			fmt.Println("Fun playing with you, shutting down...")
			os.Exit(0)
		default:
			fmt.Println("That wasn't a valid choice, please try again.")
			continue
		}

		botChoice := choices[rand.Intn(3)]

		loseCase := winCases[botChoice]
		if choice == loseCase {
			fmt.Printf("You got absolutely destroyed by the bigbrain computer, as I chose %s versus your %s. Wanna play again?\n", botChoice, choice)
		} else if choice == botChoice {
			fmt.Println("Aw man, it was a tie... Let's play again.")
		} else {
			fmt.Printf("You beat me, human. I chose %s but you chose %s... Now let's play again.\n", botChoice, choice)
		}
	}
}

func input(prompt string) string {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print(prompt)
	text, _ := reader.ReadString('\n')

	return strings.ToLower(strings.TrimSpace(text))
}
